internal/testtypes/matchers.go

```diff
@@ -0,0 +1,275 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testtypes
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/onsi/gomega/format"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// This code is a modified version of the mergePatch code from the controllers/topology/internal/mergepatch pkg.
+// TODO: Create a shared patchdiff library that can be shared for both patching and matching.
+
+// These package variables hold pre-created commonly used options that can be used to reduce the manual work involved in
+// identifying the paths that need to be compared for testing equality between objects.
+var (
+
+	// IgnoreAutogeneratedMetadata contains the paths for all the metadata fields that are commonly set by the
+	// client and APIServer. This is used as a MatchOption for situations when only user-provided metadata is relevant.
+	IgnoreAutogeneratedMetadata = IgnorePaths{
+		{"metadata", "uid"},
+		{"metadata", "generation"},
+		{"metadata", "creationTimestamp"},
+		{"metadata", "resourceVersion"},
+		{"metadata", "managedFields"},
+		{"metadata", "deletionGracePeriodSeconds"},
+		{"metadata", "deletionTimestamp"},
+		{"metadata", "ownerReferences"},
+		{"metadata", "selfLink"},
+		{"metadata", "generateName"},
+	}
+)
+
+// Matcher is a Gomega matcher used to establish equality between two Kubernetes runtime.Objects.
+type Matcher struct {
+	// original holds the object that will be used to Match.
+	original runtime.Object
+
+	// diff contains the delta between the two compared objects.
+	diff []byte
+
+	// options holds the options that identify what should and should not be matched.
+	options *MatchOptions
+}
+
+// EqualObject applies the MatchOptions and returns a Matcher for the passed Kubernetes runtime.Object. This function
+// can be used directly as a Gomega Matcher in Gomega Assertions.
+func EqualObject(original runtime.Object, opts ...MatchOption) *Matcher {
+	matchOptions := &MatchOptions{}
+	matchOptions = matchOptions.ApplyOptions(opts)
+
+	// set the allowPaths to '*' by default to not exclude any paths from the comparison.
+	if len(matchOptions.allowPaths) == 0 {
+		matchOptions.allowPaths = [][]string{{"*"}}
+	}
+	return &Matcher{
+		options:  matchOptions,
+		original: original,
+	}
+}
+
+// Match compares the object it's based on to the passed object and returns true if the objects are the same according to
+// the Matcher MatchOptions.
+func (m *Matcher) Match(other interface{}) (success bool, err error) {
```

Comment:
```suggestion
func (m *Matcher) Match(actual interface{}) (success bool, err error) {
```
I looked at a gomega matcher and this seems to be called `actual` there. I think it would be nice to do the same in our code. (same for other funcs)